ファイル用のモデルを追加

a5ob7r opened the pull request:
# 概要

オンラインストレージにおけるファイルを表現するモデルを追加する。

Fileというクラス名は、Rubyの組み込みクラスであるFileクラスと競合するため利用できない。そのため、代替としてItemというクラス名を設定する。
https://docs.ruby-lang.org/ja/latest/class/File.html

バックエンドとしてActive Storageを利用している。

a5ob7r committed:
Install active storage

a5ob7r committed:
Add Item model

a5ob7r committed:
Remove automatic file renaming when file name collisions

It's a little bit complicated to implement this feature and we don't
know which behavior is user-friendly.

Numbering position

- file(1).txt
or
- file.txt(1)

Next number

When file(1).txt, file(2).txt and file(4).txt exist, the next filename
is...

- file(3).txt
or
- file(5).txt

a5ob7r committed:
Fix the text description

The previous description is too generalized.

epaew commented on db/migrate/20230626030458_create_active_storage_tables.active_storage.rb:
```diff
@@ -0,0 +1,57 @@
+# This migration comes from active_storage (originally 20170806125915)
+class CreateActiveStorageTables < ActiveRecord::Migration[5.2]
```
`5.2` が気になる。他は `7.0` なんだよねえ。

- https://github.com/sls-training/2023-API-training-server/tree/main/db/migrate

---
このファイルって
- どこかから持ってきたものですか？
- あるいは `rails` のサブコマンドで生成したものですか？

コミットコメントまで確認したけど由来がわからんかった

epaew replied:
[Tips]

- 自分で書いたコード
- 自動生成（修正）したもの

これらのコミットは混ぜないほうが、後で git log を確認する時に便利ですね。

そして「自動生成（修正）」であれば、生成に使ったコマンド (e.g. `rails g migration user` `rubocop -a`) も書いてあるのが望ましいですねえ。

a5ob7r replied:
Railsのタスクで生成されたマイグレーションファイルです。バージョンについては、僕も気になりましたが、自動生成されたファイルだし、ActiveStorageが追加されたのはv5.2からのようだし、そういうものかなと思って特に対処しませんでした。

```shell
bin/rails active_storage:install
```

epaew replied:
私のコメントの内容は「7.0 に変えろ」ではないので、  
`rails` サブコマンドの生成結果そのままなのであれば、そこまで気にするものでもないかなとは思いました。

`[5.2]` とか `[7.0]` ってのは「Rails の migration DSL に breaking change が入っても冪等性を維持するための仕組み」だと理解しているので、「ただ 7.0 に変えるだけ」では「動く保証がないもの」だとも思ってる。

a5ob7r replied:
次のRailsのバージョンからは7.0になりそうです。

https://github.com/rails/rails/commit/506e357f00698aacf8f4d58b1c0ae26470dcdf32

a5ob7r replied:
次のというか、次のメジャーバージョンか。

epaew commented on app/models/user.rb:
```diff
@@ -2,6 +2,7 @@
 
 class User < ApplicationRecord
   has_many :access_tokens, dependent: :destroy
+  has_many :files, class_name: 'Item', dependent: :destroy
```
ここ `:files` にするのがほんとにいいのかな。`user.files[0].file` みたいな取り方をしたいのか？

かといって ActiveStorage を使ったことがないので、↓みたいなことができるのかどうかも知らないのだけれど
```ruby
has_many :items
has_many :files, through: :items
```

a5ob7r replied:
Itemモデルはリレーション用の中間モデルではなく、ActiveStorageによるファイル用のモデルのラッパーモデルなので、特にthroughを使ってアクセス方法を変更する必要はないと思います。

まあ、 `user.files[0].file` のように同じ単語がチェーンされることに関しては、気にならないこともないですが。

a5ob7r committed:
Use if and blank? instead of unless and present?

tk0miya commented on Gemfile:
```diff
@@ -7,6 +7,7 @@ ruby '3.2.2'
 
 gem 'rails', '~> 7.0.5'
 
+gem 'active_storage_validations', '~> 1.0.4'
```
メモ:
https://github.com/igorkasyanchuk/active_storage_validations
https://rubygems.org/gems/active_storage_validations/versions/1.0.4

* Star: 897
* Downloads: 900万

tk0miya commented on app/models/item.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class Item < ApplicationRecord
```
雑談? コメント?

Item という名前はかなり汎用的でほぼ意味を持たない (いろんなデータを item と呼べてしまう) ので、すごく短命の変数以外にはこの名前をつけるのには反対です。
今回はトレーニングなので深追いせずにこのまま進めるのか、リネームを考えるのかはおまかせしますが、こうした名前はあまり好まれないというのは覚えておいてほしいです。

tk0miya replied:
たぶんここでうまく名前がつけられていないので、 User から has_many 関連をつける際に `class_name` で調整が必要になってるんですよね。

a5ob7r replied:
抽象的すぎる命名だったとは思っています。

本当はFileとしたかったですが、Fileクラスが組み込みクラスに存在するので、名前空間を分離させるか、別の名前をつけるか、くらいしか方法がありませんでした。今回は後者のうち、修飾語がつかない1単語で表現できるクラス名にしたくて、結局この名称になりました。いくつか候補はありましたが、それらも抽象的すぎてアプリケーションには向かないものでしたね。現実的には、xxxFileにすることになるのかな。

a5ob7r replied:
> 修飾語がつかない1単語で表現できるクラス名にしたくて

ここは単なる趣味です。クラスの目的に対応するそのものズバリな命名にしたかったのですが、その目的からそれて抽象的な命名になってしまったのですから、世話ないですね。以後気をつけます。

tk0miya replied:
実際のシステム開発では、作る対象のシステムが複雑なことが多いので、あまりシンプルなものにはならないですね。
業務フローであったり、関連する法律であったり、要求自体が複雑なので、作るシステムも自然と複雑になっていきます。
また、時間が経過するにつれて機能が追加されることも多々あるります。

それもあって、シンプルすぎる名前については嫌な匂いを感じることがありますね。
その点では、今回 `File` が使えたとしてもそれでいいのだろうか、ともやもやしそうです。

※ 繰り返しますが、今回はあくまでトレーニングですし、異議があるわけではないですよ。


ただ、適当な抽象度が選択できなかった場合、あとあとコードの違和感が広がっていくので、それは気にしておいたほうが良さそうです。ソースコードは書く回数よりも読む回数のほうが圧倒的に多いので、まあいいかで済ませたものを今後数年読み続けることになります。
(大抵のシステムは 5年ぐらいは使われることが多いです)

tk0miya replied:
自分だったらなんて名前をつけるのだろうか、というのをしばらく考えてみたのだけど、思いついたのはこのあたりかな。

* `UploadFile`
    * アップロードされたファイル
    * どういう経緯で保存されたのかを示したもの。
    * DB に保存された時点で「アップロードされたかどうか」というのはどうでも良くなっているので、あまりデータの趣旨をちゃんと表現できているとはいえない
    * 悪くはないけど、最高の名前だとは思えないかな。
* `FileEntry`
    * ファイルエントリー。
    * ファイルシステムっぽく表現してみた。i-node 的な位置づけのデータだと考えるとそんなに違和感はない。
    * かなり汎用に寄せた名前ではあるけれど、ファイルアップローダー、ファイル管理サービスの根幹のテーブルなので、特化した名前がつけづらいし、許容できる気がする
* `UserFile`
    * ユーザーの保有しているファイル
    * 抽象度を下げて、実装にかなり寄せたネーミング
    * 特におかしくはないのだけど、もう少しいい名前がほしいな、と感じる。悪くはない。

この中から選ぶとすると `FileEntry` にしそうな気がする。
この結論を踏まえると、 `File` が使えたなら `File` を使うのはそこまで違和感はなさそう。

※ システムの目的が変わって、画像とか動画とか、そういう種類が入るともう少しネーミングの考えようがあったかもなあ、というお気持ち。

epaew replied:
私だったら
- `Upload ed File`
- `UserFile`
- `(File)Descriptor`

あたりですかねえ。

`FileEntry` も悪くはないと思います。「僕だったら思いつかない」気がしたので、箇条書きに入れるのをためらった、ぐらい

a5ob7r replied:
1単語の中で考えていたのはこんなものでした。最後の方はヤケクソでした。

- file
- item
- entry
- post
- entity
- node
- resource
- media
- record
- box
- collection

tk0miya replied:
あ"ー > ed

FileDescriptor は -1 したいですね。僕の知る限り、fd はファイルを開いたときに付く識別子のことで、保存されているファイルに使う名前ではないはず。

tk0miya replied:
>1単語の中で考えていたのはこんなものでした。最後の方はヤケクソでした。

その名前を見て、どういうデータなのかが識別できない名前は危ない気がしますね。
よほどそのシステムで重要な位置を占めていて、〜〜といえばこれのこと、と言い切れるぐらいのものでないと危ないです。
挙げてもらった名前の殆どは汎用的で意味が薄く、代名詞の it ぐらいの意味しかないので嫌な匂いがします。

resource, media あたりはシステムによっては「重要な位置を占めていて〜」というのを満たす可能性があるとは思いますが、今回のシステム向きではなさそうな気はしますね。

tk0miya replied:
実際の案件だと、バージョン2 として

* ユーザー間で共有できるようなファイルを作りたい
* Google Drive の共有ドライブみたいなもの作れない?

という話が出てくることがあって、「ぎゃー、 `File` とか `FileEntry` 、`UploadedFile` の立ち位置が困る」とかってなるんですよね。

前もってどういう要求が来るのかは予想がつかない (要求が来ないかもしれない) ので、備えておくことは難しいのですで、だからどうしよう、という話ではないのですけどね。

ただ、最初に作ったシステムが永遠に動くわけではないので、他の概念が導入されることを予期しておいた方がよいだろうな、というのはあります。
上でコメントした汎用的な名前(item, entry など)は避けた方がいいだろう、というのはこのあたりから来ています。

※ ファイル以外の、別のデータが増えたときに衝突しやすくなるのはいまいちですよね。

tk0miya commented on app/models/item.rb:
```diff
@@ -0,0 +1,19 @@
+  validates :description, length: { maximum: 1024 }
+  validates :file, attached: true, size: { less_than_or_equal_to: 1.megabytes }
+
+  # TODO: コールバック以外に適切な方法があるかもしれない。
+  before_validation :set_fallback_filename, if: -> { name.blank? }
```
コールバックでいいと思いますよ。

tk0miya commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      let(:file) { fixture_file_upload 'empty.txt' }
+
+      before do
+        described_class.new(file:, user:).save
```
こう書くのが良さそう?
```suggestion
        described_class.create!(file:, user:)
```

* new & save より create の方が一度で済むのでシンプルでは
* create より create! の方がエラーに気づきやすいのでよいのでは (もしくは save より save! の方が〜)
    * create も save もバリデーションエラーなどの際は false を返すだけなのでサイレントに失敗します

a5ob7r replied:
確かにそうですね。修正します。

tk0miya commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      let(:another_user) { FactoryBot.create :user }
+
+      before do
+        another_user.files.create file:
```
上でコメントしたように create! の方がいいですね
```suggestion
        another_user.files.create! file:
```

a5ob7r replied:
修正しました。

tk0miya commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,68 @@
+      it { is_expected.to be_falsey }
+    end
+
+    context '名前を省略したとき' do
```
この context はひとつ上かひとつ下に移動したほうがすっきりしそうです。
現状だと以下のように並んでいるんですよね。

* context '同名のファイルが存在するとき' do
* context '名前を省略したとき' do
* context '別のユーザが同名のファイル名のファイルを所有しているとき' do

同名のファイルの context は隣り合っていたほうがわかり易くないですか?

a5ob7r replied:
わかりやすいかは、なんともわかりませんが、少なくとも見栄えはそのほうがいいですね。修正します。

a5ob7r replied:
コードとしても、関心ごとが似ているのならより近くに記述されているほうが読みやすいか。

tk0miya commented on app/models/item.rb:
```diff
@@ -0,0 +1,19 @@
+
+  private
+
+  def set_fallback_filename
```
この fallback ってつけたほうがいいかなあ。
フォールバックという意味は before_validation で表現できているような気がするし、"fallback_filename" というデータがあるように錯覚してしまうので、単に `set_filename` とか `fill_filename` 、 `complement_filename` あたりがいいんじゃないかなあ。

a5ob7r replied:
> フォールバックという意味は before_validation で表現できているような気がするし、

なるほど。

a5ob7r replied:
このメソッド名で表現したかったことは、代入される値がどういったものか、ということだと気がついたので、その値を#original_filenameという関数に切り出して、値の代入はlambaで行うようにしました。

a5ob7r replied:
これくらいだったら、全部lambdaにインラインで書いてしまってもよさそうとは思いました。ただその場合、自分であればコメントで「アップロードされたファイルの元々のファイル名を設定する。」などと書きたくなるので、それならメソッド名で表現したほうがいいと判断してこのような記述にしました。

tk0miya replied:
original_filename はいいネーミングですね。新しいコードのほうがやりたいことを表現していていい感じだと思います。

a5ob7r committed:
Change how to save resources for test to .create!

These resource creations must be successful in #before, so methods with
a bang are more appropriate because they throw an exception when causes
any errors. This behavior is more observable for us about some errors.

a5ob7r committed:
Reorder test cases

Put some tests about the same named files side by side.

a5ob7r committed:
Fix a test description

Make it more detail.

a5ob7r committed:
Switch to the lambda style to fallback the filename

In this case, it's the most important to explain what the assign value
has a mean in this application. So we only need to extract the value as
a method. And the method name such as #set_xxx is a bad name for not
setter methods.

a5ob7r commented:
こちらのPRは、修正が必要と思われる箇所への対処は完了しているので、レビューをしていただけますと幸いです。

KT-tim commented on db/migrate/20230626030458_create_active_storage_tables.active_storage.rb:
```diff
@@ -0,0 +1,57 @@
+  end
+
+  private
+    def primary_and_foreign_key_types
```
privateと書いた直後はインデント下げなくても大丈夫です

KT-tim replied:
修正するとしたら別PRで大丈夫です

epaew replied:
これ、Rails のコードを読んでると出てくるスタイルなのよねえ。

「`rails` のサブコマンドで生成した」とも聞いた（ https://github.com/sls-training/2023-API-training-server/pull/39#discussion_r1243063611 ）ので、Rails リポジトリにコミットされてるテンプレートがこのスタイルだったんだと理解してます。

僕らがこのインデントスタイルで書くことはないし、RuboCop を掛けろと言われればそうだけど。
